File: BikeHistory/Models/Parts/Part.cs
namespace BikeHistory.Models.Parts;

public class Part
{
    public Guid Id { get; set; }
    public string Manufacturer { get; set; }
    public string Model { get; set; }
    public double Mileage { get; set; }
    public Guid TypeId { get; set; }
    public Guid BikeId { get; set; }

    public PartDto CreateDto()
    {
        return new PartDto(Id, Manufacturer, Model, Mileage, TypeId, BikeId);
    }

    public Part(string manufacturer, string model, double mileage, Guid typeId)
    {
        Id = Guid.NewGuid();
        Manufacturer = manufacturer;
        Model = model;
        Mileage = mileage;
        TypeId = typeId;
    }
}

File: src/WebUI/src/models/parts/part.ts
import { DataArrayResponse, FetchResponse, httpDelete, httpGetWithBody, httpPost, httpPut } from "../httpMethods.ts";

export class Part {
    id: string;
    manufacturer: string;
    model: string;
    mileage: number;
    typeId: string;
    bikeId: string;

    constructor(...args: Part[]) {
        this.id = "";
        this.manufacturer = "";
        this.model = "";
        this.mileage = 0.0;
        this.typeId = "";
        this.bikeId = "";
        if (args.length === 1) {
            Object.assign(this, args[0]);
        }
    }

    clear(): void {
        this.id = "";
        this.mileage = 0.0;
        this.model = "";
        this.manufacturer = "";
        this.typeId = "";
        this.bikeId = "";
    }

    async addPartRequest(): Promise<FetchResponse<null>> {
        return await httpPost<Part>("/api/part", this);
    }

    async deletePartRequest(): Promise<FetchResponse<null>> {
        return await httpDelete("/api/part", this.id);
    }

    async putPartRequest(): Promise<FetchResponse<null>> {
        return await httpPut<Part>("/api/part", this);
    }
}

export async function getPartsRequest(): Promise<Part[]> {
    const result = await httpGetWithBody<DataArrayResponse<Part>>("/api/part");
    if (result.status === 200) {
        return result.body.data.map(createPart);
    }
    return [];
}

/**
 * Creates a new object of the Part class from an object conforming to the IPart interface
 * @param part The object whose fields will be transferred to new object
 * @returns A new object of the Part class
 */
function createPart(part: Part) {
    return new Part(part);
}
